app/helpers/application_helper.rb

```diff
@@ -3,33 +3,34 @@ module ApplicationHelper
   include HomeHelper
 
   # Add autogenerated html data-id attribute if not present (prefixed with "aid_")
-  def link_to(*args, &block)
-    if block_given?
-      super(*args)
-    else
-      options = args[1] || {}
-      html_options = args[2] || {}
-      unless html_options.has_key?(:'data-id') || (options.is_a?(String) && options.starts_with?("mailto:"))
-        begin
-          path = URI.split(url_for(options) || html_options['href'])[5].split(/\//).select {|x| !x.empty?}
-          if !path.empty?
-            max = (path.size <= 3) ? path.size : 3
-            id = path.last(max).join('_')
-          else
-            id = 'not_defined'
-          end
-        rescue => e
-          Foreman::Logging.exception("Failed generating link using #{args.inspect}", e)
-          id = 'not_parseable'
-        end
-        html_options[:'data-id'] = "aid_#{id}"
+  def link_to(*args)
+    return super(*args) if block_given?
+
+    options = args[1] || {}
+    html_options = args[2] || {}
+    unless html_options.has_key?(:'data-id') || (options.is_a?(String) && options.starts_with?("mailto:"))
+      begin
+        # only create it once...
+        @uri_keys ||= [ :scheme, :userinfo, :host, :port, :registry, :path, :opaque, :query, :fragment ]
+        uri_elements = Hash[@uri_keys.zip(URI.split(url_for(options) || html_options['href']))]
+        path = (uri_elements[:path] || '').split('/').select {|x| !x.empty?}
```

Comment:
You can make this much simpler without relying on URI internals
```ruby
path = URI(url_for(options) || html_options['href']).path.split('/').reject(&:empty?)
```

Reply:
I tried to do two big things: solve as much as possible the access to an array element using a literal index location, and the ability to try and keep things the same as possible, so only code smells that I understand what they do will be changed.

having said that, your solution is more elegant :D